File: Reverse a Linked List/reverseALinkedList.cpp
#include <bits/stdc++.h>
using namespace std;

/* Node structure of LinkedList */

struct Node {
    
	int data;
	struct Node* next;
	
	Node(int data){
	    
		this->data = data;
		next = NULL;
	}
};

struct LinkedList {
    
	Node* head;
	LinkedList() { head = NULL; }

	
	void push(int data){
	    
		Node* temp = new Node(data);
		
		if(head == NULL ){
	        head = temp;
	    }
	    
	    else{
	        
		Node* curr = head;
		
		while( curr->next != NULL ){
		    curr =  curr->next;
		}
		
		curr->next = temp;
		
	    }
	}
	
	// Function to reverse the linked list 
	
	void reverse(){
		
		Node* curr = head;
		Node *prev = NULL, *next = NULL;

		while (curr != NULL) {
			
			next = curr->next;

			curr->next = prev;

			prev = curr;
			
			curr = next;
		}
		
		head = prev;
	}

	
	void print(){
	    
		struct Node* temp = head;
		while (temp != NULL) {
			cout << temp->data << " ";
			temp = temp->next;
		}
	}

};

int main(){
	
	LinkedList L;
	
	L.push(1);
	L.push(2);
	L.push(3);
	L.push(4);
    
	L.reverse();

	L.print();
	
	return 0;
}

File: 20_Reverse a Linked List/Ideal Solutions/reverseALinkedList.js
class Node {
  constructor(element) {
    this.element = element;
    this.next = null;
  }
}

class Linkedlist {
  constructor() {
    this.head = null;
    this.size = 0;
  }

  add(val) {
    var node = new Node(val);

    var current;

    if (this.head == null) this.head = node;
    else {
      current = this.head;

      while (current.next) {
        current = current.next;
      }

      current.next = node;
    }
    this.size++;
  }

  print() {
    if (this.head === null) return "empty";

    let curr = this.head;

    while (curr) {
      console.log(curr.element);
      curr = curr.next;
    }
  }

  reverse() {
    if (!this.head) return this.head;

    let curr = this.head;
    let prev = null;

    while (curr) {
      let tmp = curr.next;
      curr.next = prev;
      prev = curr;
      curr = tmp;
    }
    this.head = prev;
  }
}

const LL = new Linkedlist();

LL.add(3);
LL.add(2);
LL.add(1);
/* LL.print(); */
LL.reverse();
LL.print(); //1 , 2, 3
